### PR TITLE
Add the TopicInfo data structure

### LINKED ISSUE #539: Add a data structure for topic info
We're starting to copy-paste a nested dictionary with a fixed structure that represents the extracted journal/article topics, e.g.:
```python
{
    "source": "pubmed",
    "path": str(path.resolve()),
    "topics": {
        "journal": {
            "MeSH": journal_topics,
        },
        "article": {
            "MeSH": article_topics,
        },
    },
    "metadata": {
        "created-date": datetime.datetime.now().strftime(
            "%Y-%m-%d %H:%M:%S"
        ),
        "bbs-version": bluesearch.version.__version__,
        "element_in_file": i,
    },
}
```
It could be appropriate to introduce a data structure do deal with this instead.

For more context/details see this comment: https://github.com/BlueBrain/Search/pull/533#discussion_r778227276

### PR DESCRIPTION
**Pull request by Stannislav**
Fixes #539 .

## Description
This adds the `TopicInfo` data class that replaces the manual creation and copy-pasting of the structured dictionary that we have been using for this purpose so far.

## How to test?
* See new tests under `tests/unit/database/test_topic_info.py`
* Adjusted tests for old code pass

## Checklist

- [x] This PR refers to an issue from the [issue tracker](https://github.com/BlueBrain/Search/issues).
  (if it is not the case, please create an issue first).
- [x] Unit tests added.
  (if needed)
- [x] Documentation and `whatsnew.rst` updated.
  (if needed)
- [x] `setup.py` and `requirements.txt` updated with new dependencies.
  (if needed)
- [x] Type annotations added.
  (if a function is added or modified)
- [x] All CI tests pass.

### CODE REVIEW

File: src/bluesearch/database/topic_info.py
Comment (EmilieDel) [inline src/bluesearch/database/topic_info.py:108]: Maybe we can keep the absolute path here ?

Update: I just saw that you are taking the absolute path when instantiating the TopicInfo into the entrypoint. Don't you think it could be worth to still take the absolute path here to be sure it is harmonized ?

File: src/bluesearch/database/topic_info.py
Comment (EmilieDel) [inline src/bluesearch/database/topic_info.py:104]: Just for me to understand, what do we expect inside `self.metadata` ? 
Shouldn't `element_in_file` be part of the metadata  ?

File: src/bluesearch/database/topic_info.py
Comment (EmilieDel) [inline src/bluesearch/database/topic_info.py:111]: 1. Just for me to understand, why do you think it is needed to do a `deepcopy` here ? 
2. Just to put into light, implementing the `topics` value this way means that we will have an `article` and `journal` field even when we do not search for one of them. Do you think it is a problem ?

**Review by EmilieDel — APPROVED** [review]
Really nice 🎉  It is definitely more clear and clean than copy-pasting the dictionary for every source, thanks for the changes!

File: src/bluesearch/database/topic_info.py
Comment (Stannislav) [reply src/bluesearch/database/topic_info.py:111]: 1. It's to make sure you can't mutate your `TopicInfo` by changing the returned dictionary. Without deepcopy if your do `json = topic_info.json(); json["topic"]["article"]["x"] = 5` then this will also change the internal state of `topic_info`.
2. Yes, this is actually a slightly different behaviour to what we used to do. Do you know if we rely on the presence/abcense of the "article"/"journal" keys? I couldn't see it specified anywhere. I reconed it would be easier to deal with if it's always present and of the same type, so you don't have to constantly check for Nones.

File: src/bluesearch/database/topic_info.py
Comment (Stannislav) [reply src/bluesearch/database/topic_info.py:108]: That's certainly possible. I left the responsibility for checking/resolving the path to the caller for two reasons
* We accept `pathlib.Path` and `str` types, so we'd have to convert back and forth
* If the path doesn't exist then resolving it could give unexpected results (from the [docs](https://docs.python.org/3/library/pathlib.html#pathlib.Path.resolve): "[if] the path is resolved as far as possible and any remainder is appended without checking whether it exists")

But it's not an absolute no-go. What do the others think?

File: src/bluesearch/database/topic_info.py
Comment (Stannislav) [reply src/bluesearch/database/topic_info.py:104]: Right now it's just possible to pass any additional metadata via `TopicInfo(..., metadata={"some key": "some value", ...})`, that's what will be in `self.metadata`. But if we don't want that, I guess we can get rid of that. I'm actually not so sure here about what we want exactly.

I'm not sure I understand the question about `element_in_file`? It's inserted into `metadata` if it's set, otherwise it will be absent. At least that's how it was specified in https://github.com/BlueBrain/Search/issues/518#issuecomment-985525160.

File: src/bluesearch/database/topic_info.py
Comment (jankrepl) [inline src/bluesearch/database/topic_info.py:50]: Maybe we can name it `_add_topics` to stress that it will only be used internally.

Completely subjective and feel free to ignore

File: src/bluesearch/database/topic_info.py
Comment (jankrepl) [inline src/bluesearch/database/topic_info.py:102]: 1. Maybe we can document somewhere (class docstring?) that the `created-date` and `bbs-version` will be automatically added?
2. If the user is not aware of 1. they might accidentally provide `created-date` and `bbs-version` manually. Using the `{}` literal actually silently allows for duplicate keys. Maybe we can use `dict` instead to prevent this from happening (it raises an exception)

```python
In [13]: d_1 = {"a": 2}

In [14]: d_2 = {**d_1, "a": 2323}

In [15]: d_2
Out[15]: {'a': 2323}

In [16]: d_3 = dict(**d_1, a=2323)
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-16-6dfbfab8a8f2> in <module>
----> 1 d_3 = dict(**d_1, a = 2323)

TypeError: type object got multiple values for keyword argument 'a'
```

But whatever, it is a detail.

File: src/bluesearch/database/topic_info.py
Comment (jankrepl) [inline src/bluesearch/database/topic_info.py:43]: Might be nice to document these in the class docstring under `Attributes` (I guess that is the right section).

File: src/bluesearch/database/topic_info.py
Comment (jankrepl) [reply src/bluesearch/database/topic_info.py:108]: IMO both are okay as long as we clearly document the behavior somewhere (e.g. class docstring).

**Review by jankrepl — COMMENTED** [review]
Very reasonable! Thank you for the effort:)

**Comment by FrancescoCasalegno** [comment]
### Discussed 2022-01-11

Let's wait for #535 and #540 to be merged into `master` before merging this PR.
Notice that the code will need to be changed to address the updates on the `master`.

File: src/bluesearch/database/topic_info.py
Comment (EmilieDel) [reply src/bluesearch/database/topic_info.py:104]: Sorry, I am realising that my comment was not really clear and maybe not at the right place. I was wondering if `element_in_file` should be a separate attribute or if we should expect the user to give this info through the metadata dictionary (as it is really source specific and going to end in the metadata field anyway) but maybe it is better to keep it explicit.

File: src/bluesearch/database/topic_info.py
Comment (EmilieDel) [reply src/bluesearch/database/topic_info.py:111]: 1. I see, makes sense. Thanks!
2. I don't think we currently directly rely on this. I thought it was going to be though to distinguish when we did not look for a topic and when looked for it but did not find any. But in the first case, the dictionary is going to be empty and in the second one, the dictionary is going to have a key with the value being an empty list. So maybe it is not really a problem. Moreover, we can still check the bbs code version.

File: src/bluesearch/database/topic_info.py
Comment (FrancescoCasalegno) [reply src/bluesearch/database/topic_info.py:108]: I feel like Emilie that absolute paths are better, since we are then expecting to use this info when reading the file to the know which files should be loaded into the DB and which ones should not.

File: src/bluesearch/entrypoint/database/topic_extract.py
Comment (FrancescoCasalegno) [inline src/bluesearch/entrypoint/database/topic_extract.py:153]: I think it would be cleaner to handle at arg-parsing time this casting of the `str` passed on the command line into an `ArticleSource`.

This means the following small changes:
1. Specify the `type` in the `ArgumentParser`:
    ```python
    parser.add_argument(
        "source",
        type=ArticleSource,
        choices=(
            "arxiv",
            "biorxiv",
            "medrxiv",
            "pmc",
            "pubmed",
        ),
        choices=[member.value for member in ArticleSource.__members__.values()],
    ```
3. Change the signature of `run()` to take an `ArticleSource`:
    ```python
   def run(
      *,
      source: ArticleSource,
      input_path: Path,
      output_file: Path,
      match_filename: str | None,
      recursive: bool,
      overwrite: bool,
      dry_run: bool,
     ) -> int:
   ```
3. Remove the then unnecessary `try... except` block here.

File: src/bluesearch/database/topic_info.py
Comment (FrancescoCasalegno) [inline src/bluesearch/database/topic_info.py:90]: Wouldn't it make sense to leverage [`dataclasses.asdict`](https://docs.python.org/3/library/dataclasses.html#dataclasses.asdict) here?

**Review by FrancescoCasalegno — COMMENTED** [review]
Thank you for your work @Stannislav Stan! 

It will definitely make our code simpler to maintain and understand!

File: src/bluesearch/entrypoint/database/topic_extract.py
Comment (Stannislav) [reply src/bluesearch/entrypoint/database/topic_extract.py]: Cool, thanks! Commited in cba25d0

File: src/bluesearch/entrypoint/database/topic_extract.py
Comment (Stannislav) [reply src/bluesearch/entrypoint/database/topic_extract.py:153]: Yes, makes total sense. However, it seems that for some reason `argparse` does not play well with enums :( The use of enums is explicitly discouraged [in the docs](https://docs.python.org/3/library/argparse.html#choices). There's also a discussion about it [in the bug tracker](https://bugs.python.org/issue25061).

Long story short, this:
```python
parser.add_argument(
    "source",
    type=ArticleSource,
    choices=[member.name for member in ArticleSource],
```
doesn't work because argparse first applies the callable in the `type` argument and only then checks against the `choices`, which never matches because it compares `str` to `ArticleSource`.

This:
```python
parser.add_argument(
    "source",
    type=ArticleSource,
    choices=ArticleSource,
```
works, but the help message is misleading:
```
 bbs_database topic-extract --help
usage: bbs_database topic-extract [-h] [-v] [-m MATCH_FILENAME] [-R] [-o] [-n]
                                  {ArticleSource.ARXIV,ArticleSource.BIORXIV,ArticleSource.MEDRXIV,ArticleSource.PMC,ArticleSource.PUBMED,ArticleSource.UNKNOWN}
                                  input_path output_file
```
So one might have to stick with `str` type and casting to `ArticleSource` later. The `try-except` is probably not necessary if we use `choices`.

Comment (Stannislav) [reply src/bluesearch/entrypoint/database/topic_extract.py]: I removed the `try-except` in b1df1ae (+some cleaning up), but didn't didn't add `type=ArticleSource` otherwise.

File: src/bluesearch/database/topic_info.py
Comment (Stannislav) [reply src/bluesearch/database/topic_info.py:90]: I'm not sure how we'd do it in our case since we return a dict with a different struture and also add additional values (creation date and bluesearch version) that are not members of the dataclass itself. Do you have a suggestion?

File: src/bluesearch/database/topic_info.py
Comment (Stannislav) [reply src/bluesearch/database/topic_info.py:43]: I don't like this metadata either. I was following the spec here https://github.com/BlueBrain/Search/issues/518#issuecomment-985525160 where the idea of metadata was to contain any source-specific information.

I can actually remove it now because we don't have anything source-specific at the moment. Note though that the `metadata` will still be part of the dict returned by `TopicInfo.json()` and will contain a dictionay with `created-date`, `bbs-version`, and an optional `element_in_file` since this is what we put in there [on master](https://github.com/BlueBrain/Search/blob/d41611ad516e05b2fe28007659b4f33f3d6ce1a1/src/bluesearch/entrypoint/database/topic_extract.py) by hand.

Comment (Stannislav) [reply src/bluesearch/database/topic_info.py]: Removed the `metadata` parameter in 827c79d

File: src/bluesearch/database/topic_info.py
Comment (Stannislav) [reply src/bluesearch/database/topic_info.py]: Ah I see :) There's actually a discussion thread with @FrancescoCasalegno about the `metadata` parameter further below. For now I completely removed it in 827c79d. So I guess for now it has to stay as a separate paramter.

Comment (Stannislav) [reply src/bluesearch/database/topic_info.py]: But feel free to chime in on the structure discussion below, I think we're still looking for a good solution :)

File: src/bluesearch/entrypoint/database/topic_extract.py
Comment (Stannislav) [reply src/bluesearch/entrypoint/database/topic_extract.py]: Good idea. Added in fb0d21f

File: src/bluesearch/database/topic_info.py
Comment (Stannislav) [reply src/bluesearch/database/topic_info.py]: Sure, why not! Changed in 0ec90d4

File: src/bluesearch/database/topic_info.py
Comment (Stannislav) [reply src/bluesearch/database/topic_info.py]: Actually, this `metadata` parameter got removed in 827c79d following an ongoing discussion with @FrancescoCasalegno below, so I guess this solves 2.

File: src/bluesearch/database/topic_info.py
Comment (Stannislav) [reply src/bluesearch/database/topic_info.py]: Just tried and learned that `sphinx`/`autodoc`/`napoleon` don't play well with dataclasses. They're not clever enough to recognise them and document everything as class variables. If one includes a class docstring with `Parameters` and `Attributes` then sphinx generates duplicate entries and even issues warnings for `Attributes` because they are already implicitly included in the doc.

I haven't found a way to override this behaviour, there's a very ugly way of inline-documenting attributes/class variables:
```python
    source: ArticleSource
    path: str | pathlib.Path
    element_in_file: int | None = None
    #: A mapping of the form topic-kind -> list-of-topics that contains
    #: article topics. For example: ``{"MeSH": ["topic 1", "topic 2"]}``.
    #: Use ``add_article_topics`` method to populate it.
    article_topics: dict[str, list[str]] = field(init=False, default_factory=dict)
    #: A mapping of the form topic-kind -> list-of-topics that contains
    #: journal topics. For example: ``{"MeSH": ["topic 1", "topic 2"]}``.
    #: Use ``add_journal_topics`` method to populate it.
    journal_topics: dict[str, list[str]] = field(init=False, default_factory=dict)
```
but to my eyes this looks extremely ugly.

I can think of two ways of going forward:
* Not documenting attributes of dataclasses at all
* Documenting paramters in the class docstring (but they will be duplicated as class variables because that's what sphinx thinks they are) and attributes inline like in the example above.

Any suggestions?

File: src/bluesearch/database/topic_info.py
Comment (Stannislav) [reply src/bluesearch/database/topic_info.py]: Regarding 1. is the change in c9fc877  what you had in mind?

File: src/bluesearch/database/topic_info.py
Comment (Stannislav) [reply src/bluesearch/database/topic_info.py:111]: Path resolution added here: 5145e04

File: src/bluesearch/database/topic_info.py
Comment (Stannislav) [reply src/bluesearch/database/topic_info.py]: Actually, bad news, we can't make it private by adding an underscore:
```
src/bluesearch/database/topic_info.py:50:89: E501 line too long (89 > 88 characters)
```
🤣

File: src/bluesearch/database/topic_info.py
Comment (EmilieDel) [reply src/bluesearch/database/topic_info.py:111]: Thank you! :)

File: src/bluesearch/entrypoint/database/topic_extract.py
Comment (jankrepl) [reply src/bluesearch/entrypoint/database/topic_extract.py]: Great! Thank you!

File: src/bluesearch/database/topic_info.py
Comment (jankrepl) [reply src/bluesearch/database/topic_info.py]: 2. OK:)
1. Perfect, thanks!

File: src/bluesearch/database/topic_info.py
Comment (jankrepl) [reply src/bluesearch/database/topic_info.py]: Pretty weird. I did not know about this issue. Yeh, maybe we don't have to write any docstrings. As you want:) Feel free to resolve without any action.

Comment (Stannislav) [reply src/bluesearch/database/topic_info.py]: I'd leave it like it is for now. I think it relatively transparent right now. Maybe if we add more stuff to this class (cf. the metadata discussion) then it might be worth adding more documentation about it too.

File: src/bluesearch/entrypoint/database/topic_extract.py
Comment (FrancescoCasalegno) [reply src/bluesearch/entrypoint/database/topic_extract.py]: Thank you for the explanation, I had tried it on my laptop and I did not notice that you indeed get misleading help messages for `choices=ArticleSource`.

I tried setting `ArticleSource.__repr__` but it also does not work! 🤷

File: src/bluesearch/database/topic_info.py
Comment (FrancescoCasalegno) [reply src/bluesearch/database/topic_info.py:92]: True, even if we got rid of the `metadata` we would still have slightly different data structure in the `json` we want to return.

In fact I was thinking whether there was any real advantage in storing info in a different way in this class and in the `json` we need to use elsewhere. If we had the same structure in those two places then `asdict` would do and would streamline things imo.

But looks good to me as it is as well!

**Review by FrancescoCasalegno — APPROVED** [review]
Lgtm!

**Pull request merged**
